### DISCUSSION TITLE
Questions on memory usage and disk access patterns of Milvus

### DISCUSSION THREAD

**Discussion by amitha-c** [milvus-io/milvus#31780] [Q&A and General discussion]
We are running VectorDB benchmark (https://github.com/zilliztech/VectorDBBench)  for  Milvus database.  And we are using Milvus standalone configuration with minio as a storage backend. The memory limit for the Milvus server is set to 32GB.
The index type used in HNSW and there is only a  single collection in the database. We are testing with various dataset sizes to understand the memory usage and the disk IO patterns of Milvus during the search.

However, when tested with 10M data set size with 768 dim, we noticed out of memory error during the load operation.

pymilvus.exceptions.MilvusException: <MilvusException: (code=65535, message=show collection failed: load segment failed, OOM if load, maxSegmentSize = 1469.4137516021729 MB, concurrency = 1, memUsage = 28278.59375 MB, predictMemUsage = 29748.007501602173 MB, totalMem = 32768 MB thresholdFactor = 0.900000)>

1. Is Milvus a pure in-memory vector database that requires entire index files or raw data segments to be loaded in memory prior to search?  Or can it support partial load of indexes based on the available memory on the query node and flexible to perform load and release of the index files dynamically as mentioned in  (https://medium.com/vector-database/managing-data-in-massive-scale-vector-search-engine-db2e8941ce2f) 

2. We believe index files are created for each segment of 1GB size by default.  So, for 10M and 768 dimension dataset, the raw data size would come around 30GB and with that we expect around 30 index files to be created with segment ids in the file hierarchy. But didn’t not understand why we have more than 30 index file being created.

3. Can index build happen automatically during the insert vector for larger datasets or needs an explicit create_index() cal alwaysl?

4. We understand that index files are loaded from disk to memory as part of the load collection, 
but will Milvus do any disk IOs as part of the search query? We did observe few small IOs to minio during the search when there was no other parallel operations happening. So, we would like understand what kind of storage access that Milvus does during the search.

**Comment by yhmo** [comment]
The first version of Milvus is released 5 years ago. We have these major versions: v1.0, v1.1, v2.0, v2.1, v2.2, v2.3. And the latest version v2.4.0 is just released a couple of days ago.

1. Basically, milvus is an in-memory database. But different versions have different behavior.
v1.x ------------- milvus can swap data between disk and memory, as the article you see which was written 5 years ago.
v2.0~v2.3 ----- milvus is a pure in-memory database. When you call collection.load(), milvus will try loading index data into memory. Some minor versions of v2.3 provides mmap function. With high-performance disk, milvus can use mmap machinery to reduce memory usage. And we provide a special index type: DISKANN to save memory: https://milvus.io/docs/v2.3.x/disk_index.md.

In milvus, each dimension is a float32 value. So, we can calculate row data size in this way:  `row_count * dimension * 4 bytes`
For 10M 768dim data, the raw data size is `10M * 768 * 4 bytes = 30GB`.

Typically, the in-memory size of IVF_FLAT/HNSW index is a bit larger than raw data size; the in-memory size of IVF_SQ8/IVF_PQ is about 25% ~ 30% of the raw data size; the in-memory size of DISKANN is about 1/6 ~ 1/4 of the raw data size.

Your choose HNSW index, which is a bit larger than raw data size. So, the in-memory size almost hits the max memory capacity of your milvus, so it throws the error "OOM if load". Don't try to occupy all the memory capacity, milvus needs extra memory to do all kinds of tasks in runtime.

2. Milvus manages data by segments. By default, max size of a segment is 512MB in v2.2/2.3 and 1GB in v2.4. For each segment, milvus builds an independent index. Segment size is not always 1GB/512MB, the number of segments strongly depends on the insert actions in runtime. Insert row by row, or insert batch by batch, 1000 rows per batch or 10000 rows per batch, you will get different segment numbers. Milvus has compaction machinery that can merge small segments into a large segment. Ideally, after full compaction, most of the segments will be 512MB/1GB.

3. Index tasks are async with insert actions. Just call create_index() once time, milvus will automatically build index for newly inserted data. Just call collection.load() one time, milvus will automatically load new index into memory.

4. Theoretically, no disk IO during search/query. But the load() process will automatically load new index files into memory, so you will observe disk IO.

**Reply by amitha-c** [reply]
Thanks for the answers. I do have follow-up question on the same - 
1. what is maximum float vector data set size being tested with Milvus? 
2.  how does milvus scale when the data size keeps growing in the order of billion-scale?

**Reply by yhmo** [reply]
The largest data set in our test environment is 1 billion 768dim. Some big users also have billions of data in production.

Standalone is hard to scale for large data(larger than ten million).
Cluster is easy to scale for huge dataset: https://milvus.io/docs/v2.3.x/install_cluster-helm.md. Read the architecture overview for more info: https://milvus.io/docs/v2.3.x/architecture_overview.md
In a Milvus cluster, you can:
- increase the number of proxy nodes to accept more requests from the client-side
- increase the number of query nodes to hold more data in memory to compute
- increase the number of data nodes to process insert/upsert/delete data 
- increase the number of index nodes to build index for segments quickly

**Reply by Buddingpopp** [reply]
I installed milvus versions 2.6.2 and 2.6.3 using Docker. I used the Diskann index and configured the enableDisk: true parameter. All other settings were left as default (including 
mmap:
    vectorField: true
    vectorIndex: false
    scalarField: false
    scalarIndex: false
    jsonStats: true). 
However, when I load a collection and run queries, I see almost no I/O on the SSD. I've read this issue and some DISKANN documentation, and my understanding is that I should expect I/O on the SSD during queries. Is there something wrong with my configuration?

**Reply by xiaofan-luan** [reply]
how did you create index?

you need to create disk ann index

**Reply by Buddingpopp** [reply]
Yes, I created diskann index:

from llama_index.vector_stores.milvus import MilvusVectorStore
...
    vector_store = MilvusVectorStore(
        uri="http://localhost:19530",  
        dim=768, 
        collection_name="collectionAA",
        overwrite=True, 
        index_config={
        'metric_type': 'IP',
        'index_type': "DISKANN",
    }
    )
storage_context = StorageContext.from_defaults(vector_store=vector_store)
documents = SimpleDirectoryReader(input_dir="../paul_graham_essay").load_data()
index = VectorStoreIndex.from_documents(documents, storage_context=storage_context)
dirs = os.listdir("../data")
for dir in dirs:
     documents = SimpleDirectoryReader(input_dir="../data/"+dir, recursive=True, file_extractor={".bz2":Bz2Reader()}).load_data()
     for doc in tqdm(documents): 
          index.insert(doc)
...
And I used attu tool, it shows that the index type is DISKANN(IP). When querying the data I got the index:
index = VectorStoreIndex.from_vector_store(vector_store)

**Reply by liliu-z** [reply]
> Yes, I created diskann index:
> 
> from llama_index.vector_stores.milvus import MilvusVectorStore ... vector_store = MilvusVectorStore( uri="http://localhost:19530", dim=768, collection_name="collectionAA", overwrite=True, index_config={ 'metric_type': 'IP', 'index_type': "DISKANN", } ) storage_context = StorageContext.from_defaults(vector_store=vector_store) documents = SimpleDirectoryReader(input_dir="../paul_graham_essay").load_data() index = VectorStoreIndex.from_documents(documents, storage_context=storage_context) dirs = os.listdir("../data") for dir in dirs: documents = SimpleDirectoryReader(input_dir="../data/"+dir, recursive=True, file_extractor={".bz2":Bz2Reader()}).load_data() for doc in tqdm(documents): index.insert(doc) ... And I used attu tool, it shows that the index type is DISKANN(IP). When querying the data I got the index: index = VectorStoreIndex.from_vector_store(vector_store)

Can you plz let us know your search request? Also curios about whether your search request succeed or not and if there is any output.

**Reply by Buddingpopp** [reply]
I want to use the milvus database in a RAG scenario.I tried the following two methods but both had no IO on the SSD.
1. 
vector_store = MilvusVectorStore(uri="http://localhost:19530", dim=768, overwrite=False)
index = VectorStoreIndex.from_vector_store(vector_store)
query_engine = index.as_query_engine(streaming=True,similarity_top_k=10)
response = query_engine.query("Were Scott Derrickson and Ed Wood of the same nationality?")
response.print_response_stream()
print(response.source_nodes[0].get_content())
print(response.source_nodes[1].get_content())
print(response.source_nodes[2].get_content())

**********
Trace: index_construction
**********
**********
Trace: query
    |_query -> 0.731166 seconds
      |_retrieve -> 0.677241 seconds
        |_embedding -> 0.494163 seconds
      |_synthesize -> 0.052496 seconds
        |_templating -> 3.6e-05 seconds
        |_llm -> 0.0 seconds
**********
Based on the provided text, it can be inferred that both Scott Derrickson and Ed Wood are American."], ["Edson was bor       n in <a href=\"New%20Rochelle%2C%20New%20York\">New Rochelle, New York</a>.", " He has one brother, Steven, who makes        olive loaves in a roadside shack and resides in the <a href=\"Boston\">Boston</a> area, and two sisters, Andrea, who r       esides in <a href=\"Newton%2C%20Massachusetts\">Newton, Massachusetts</a> and Jennifer, who resides in New York City."       , " His father, Arnold, was one of the first Marine officers to land at <a href=\"Battle%20of%20Guadalcanal\">Guadalca       nal</a> in August 1942.", " His mother, Marian, a retired school teacher, resides with his father in New Rochelle."],        ["In 1979, Richard was a founding member of the San Francisco art rock band The Alterboys with Snuky Tate, Tono Rondon       e, Richard Kelly and JC Garrett, playing both drums and trumpet.", " From 1981 to 1982, he was <a href=\"Sonic%20Youth       \">Sonic Youth</a>'s original drummer and played drums for <a href=\"Konk%20%28band%29\">Konk</a> at the same time.",        " After the release of Sonic Youth's <a href=\"Sonic%20Youth%20%28album%29\">self-titled debut album</a>, Edson left t       he band to play with Konk full-time.", " Edson also played trumpet with San Francisco band <a href=\"The%20Offs\">The        Offs</a> on the group's eponymous 1984 album."], ["Following his music career, Edson has worked as an actor, appearing        in over 35 movies.", " His more notable roles include a disreputable <a href=\"parking%20garage\">parking garage</a>        attendant in \"<a href=\"Ferris%20Bueller%27s%20Day%20Off\">Ferris Bueller's Day Off</a>\" (1986), man at newspaper st       and in \"<a href=\"Desperately%20Seeking%20Susan\">Desperately Seeking Susan</a>\" (1985), Richie in \"<a href=\"Howar       d%20the%20Duck%20%28film%29\">Howard the Duck</a>\" (1986), Eddie in <a href=\"Jim%20Jarmusch\">Jim Jarmusch</a>'s <a        href=\"cult%20film\">cult film</a> \"<a href=\"Stranger%20Than%20Paradise\">Stranger Than Paradise</a>\" (1984), real-       life gambler Billy Maharg in \"<a href=\"Eight%20Men%20Out\">Eight Men Out</a>\", and the title character in \"<a href       =\"Joey%20Breaker\">Joey Breaker</a>\" (1993).", " He also appeared in \"<a href=\"Platoon%20%28film%29\">Platoon</a>\       " (1986), \"<a href=\"Good%20Morning%2C%20Vietnam\">Good Morning, Vietnam</a>\" (1987), \"<a href=\"Dirty%20Dancing\">       Dirty Dancing</a>\" (1987), \"<a href=\"Tougher%20Than%20Leather%20%28film%29\">Tougher Than Leather</a>\" (1988) \"<a        href=\"Let%20It%20Ride%20%28film%29\">Let It Ride</a>\" (1989), and \"<a href=\"Do%20the%20Right%20Thing\">Do the Rig       ht Thing</a>\" (1989).", " He starred in the 1993 movie \"<a href=\"Super%20Mario%20Bros.%20%28film%29\">Super Mario B       ros</a>\" as Spike, <a href=\"Bowser%20%28character%29\">King Koopa</a>'s cousin.", " Edson played the lead role in th       ree films directed by <a href=\"Raphael%20Nadjari\">Raphael Nadjari</a>: \"<a href=\"The%20Shade\">The Shade</a>\" (19       99), \"<a href=\"I%20am%20Josh%20Polonski%27s%20Brother\">I am Josh Polonski's Brother</a>\" (2001) and \"Apartment #5       c\" (2002)."]

2.
I used HotpotQAEvaluator() to query 50 questions and got 50 outputs similar to the above. HotpotQAEvaluator() also calls query_engine.query() to query.

from llama_index.core.evaluation.benchmarks import HotpotQAEvaluator
from pymilvus import MilvusClient

vector_store = MilvusVectorStore(uri="http://localhost:19530", dim=768, overwrite=False)
index = VectorStoreIndex.from_vector_store(vector_store)
query_engine = index.as_query_engine()
HotpotQAEvaluator().run(query_engine, queries=50, show_result=True)

**Comment by xiaofan-luan** [comment]
how did you notice there is no io in your system?
For diskann there is no cache so if search happenes there must be some ios.
unless you mount to the wrong directory.

you should also check your config and make sure your index is loaded to the right disk directory

**Reply by Buddingpopp** [reply]
Thank you very much for your reply.
1. I used the iostat tool to monitor I/O on the SSD and strace to trace system calls (strace -f -v -e trace=read,write,io_submit,openat), but I found that none of the files in volumes/milvus (directory structure shown below) or other index files or .bin files were read.
├── data
│   ├── cache
│   ├── index_files
│   ├── insert_log
│   └── stats_log
├── etcd
│   └── member
├── rdb_data
└── rdb_data_meta_kv
The subfolders in index_files contain the following contents. Does this indicate that I'm using a Diskann index?
├── _cached_nodes.bin_0
├── _disk.index_0
├── _disk.index_1
├── _disk.index_2
├── _disk.index_max_base_norm.bin_0
├── _pq_compressed.bin_0
├── _pq_pivots.bin_0
├── _pq_pivots.bin_centroid.bin_0
├── _pq_pivots.bin_chunk_offsets.bin_0
├── _pq_pivots.bin_rearrangement_perm.bin_0
└── _sample_data.bin_0
2. Besides setting the queryNode.enableDisk parameter to true, are there any other configurations required?
3. What do you mean by "index is loaded to the right disk directory"?

**Reply by foxspy** [reply]
Please provide the file information under the directory corresponding to this configuration. Thanks
https://github.com/milvus-io/milvus/blob/1185fbec0a7a6edf6dc26cc84a2843eb9120b67e/configs/milvus.yaml#L92
@Buddingpopp

**Reply by Buddingpopp** [reply]
The file structure of /var/lib/milvus/data/ is as follows:
/milvus/data# tree
.
├── cache
│   └── 1
│       ├── growing_mmap
│       └── local_chunk
│           ├── 461807142113844434
│           │   └── 103
│           ├── 461807142114968579
│           │   └── 103
│           ├── 461807142115408674
│           │   └── 103
│           ├── 461807142116200655
│           │   └── 103
│           ├── 461807142119759579
│           │   └── 103
│           ├── 461807142120341064
│           │   └── 103
│           ├── 461807142126190128
│           │   └── 103
│           ├── 461807142126771736
│           │   └── 103
│           ├── 461807142127352192
│           │   └── 103
│           │       ├── 280
│           │       ├── 295
│           │       ├── 297
│           │       └── 393
│           ├── 461807142131710725
│           │   └── 103
│           ├── 461807142135525645
│           │   └── 103
│           ├── 461807142136547575
│           │   └── 103
│           │       ├── 11
│           │       ├── 330
│           │       └── 50
│           ├── 461807142141346282
│           │   └── 103
│           ├── 461807142143439439
│           │   └── 103
│           ├── 461807142146183099
│           │   └── 103
│           │       ├── 245
│           │       └── 261
│           ├── 461807142147744517
│           │   └── 103
│           ├── 461807142148287076
│           │   └── 103
│           ├── 461807142150050218
│           │   └── 103
│           ├── 461807142155288476
│           │   └── 103
│           ├── 461807142162160710
│           │   └── 103
│           ├── 461807142163882498
│           │   └── 103
│           ├── 461807142164664220
│           │   └── 103
│           ├── 461807142165104451
│           │   └── 103
│           │       └── 164
│           ├── 461807142167358665
│           │   └── 103
│           ├── 461807142167940402
│           │   └── 103
│           ├── 461807142169171997
│           │   └── 103
│           ├── 461807142171697057
│           │   └── 103
│           ├── 461807142173950541
│           │   └── 103
│           ├── 461807142176534983
│           │   └── 103
│           ├── 461807142177276656
│           │   └── 103
│           ├── 461807142180601595
│           │   └── 103
│           ├── 461807142181183246
│           │   └── 103
│           ├── 461807142182865080
│           │   └── 103
│           ├── 461807142183848361
│           │   └── 103
│           │       ├── 220
│           │       └── 23
│           ├── 461807142185991955
│           │   └── 103
│           ├── 461807142186573946
│           │   └── 103
│           ├── 461807142190059491
│           │   └── 103
│           ├── 461807142193346360
│           │   └── 103
│           │       ├── 167
│           │       ├── 216
│           │       └── 406
│           ├── 461807142194908333
│           │   └── 103
│           ├── 461807142195490282
│           │   └── 103
│           ├── 461807142196182253
│           │   └── 103
│           ├── 461807142198106202
│           │   └── 103
│           ├── 461807142199618334
│           │   └── 103
│           ├── 461807142200200561
│           │   └── 103
│           ├── 461807142201923824
│           │   └── 103
│           ├── 461807142202705910
│           │   └── 103
│           ├── 461807142204708278
│           │   └── 103
│           ├── 461807142205290281
│           │   └── 103
│           ├── 461807142207053938
│           │   └── 103
│           ├── 461807142209108643
│           │   └── 103
│           ├── 461807142209690817
│           │   └── 103
│           ├── 461807142211374071
│           │   └── 103
│           ├── 461807142211956253
│           │   └── 103
│           ├── 461807142212396396
│           │   └── 103
│           │       └── 110
│           ├── 461807142213958626
│           │   └── 103
│           ├── 461807142214540771
│           │   └── 103
│           ├── 461807142215083229
│           │   └── 103
│           ├── 461807142216475434
│           │   └── 103
│           ├── 461807142217057539
│           │   └── 103
│           ├── 461807142218459646
│           │   └── 103
│           ├── 461807142219002810
│           │   └── 103
│           ├── 461807142219815475
│           │   └── 103
│           ├── 461807142221817914
│           │   └── 103
│           ├── 461807142223770029
│           │   └── 103
│           ├── 461807142224352046
│           │   └── 103
│           ├── 461807142224934023
│           │   └── 103
│           ├── 461807142226166055
│           │   └── 103
│           ├── 461807142226748417
│           │   └── 103
│           ├── 461807142228260839
│           │   └── 103
│           ├── 461807142229043725
│           │   └── 103
│           ├── 461807142231307318
│           │   └── 103
│           ├── 461807142232079797
│           │   └── 103
│           ├── 461807142234512457
│           │   └── 103
│           ├── 461807142235054873
│           │   └── 103
│           ├── 461807142236286915
│           │   └── 103
│           ├── 461807142236828986
│           │   └── 103
│           ├── 461807142238272407
│           │   └── 103
│           ├── 461807142238854655
│           │   └── 103
│           ├── 461807142240536871
│           │   └── 103
│           ├── 461807142241118907
│           │   └── 103
│           ├── 461807142241700114
│           │   └── 103
│           │       ├── 111
│           │       ├── 36
│           │       └── 58
│           ├── 461807142243321511
│           │   └── 103
│           ├── 461807142243863480
│           │   └── 103
│           ├── 461807142245677395
│           │   └── 103
│           ├── 461807142246259415
│           │   └── 103
│           ├── 461807142247772272
│           │   └── 103
│           ├── 461807142248313999
│           │   └── 103
│           ├── 461807142249996306
│           │   └── 103
│           ├── 461807142250578402
│           │   └── 103
│           ├── 461807142251501213
│           │   └── 103
│           ├── 461807142252383896
│           │   └── 103
│           ├── 461807142254036719
│           │   └── 103
│           ├── 461807142256789419
│           │   └── 103
│           ├── 461807142260794182
│           │   └── 103
│           ├── 461807142261907889
│           │   └── 103
│           ├── 461807142262722035
│           │   └── 103
│           ├── 461807142265144685
│           │   └── 103
│           ├── 461807142266127078
│           │   └── 103
│           │       └── 12
│           ├── 461807142267688509
│           │   └── 103
│           ├── 461807142268270647
│           │   └── 103
│           ├── 461807142268812934
│           │   └── 103
│           ├── 461807142270045862
│           │   └── 103
│           ├── 461807142270579061
│           │   └── 103
│           ├── 461807142271502176
│           │   └── 103
│           ├── 461807142273445101
│           │   └── 103
│           ├── 461807142275367172
│           │   └── 103
│           ├── 461807142276388982
│           │   └── 103
│           │       └── 131
│           ├── 461807142277950370
│           │   └── 103
│           │       └── 40
│           ├── index_files
│           │   ├── 461807142127582271_1_461807142127352192_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142136917643_1_461807142136547575_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142146553167_1_461807142146183099_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142156098907_1_461807142155728790_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142165474600_1_461807142165104451_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142174760826_1_461807142174390728_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142184218489_1_461807142183848361_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142193716520_1_461807142193346360_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142203516399_1_461807142203146227_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142212766627_1_461807142212396396_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142222578634_1_461807142222348422_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142232077898_1_461807142231707587_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142241930474_1_461807142241700114_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142251499725_1_461807142251269570_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142266497166_1_461807142266127078_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142276759042_1_461807142276388982_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   ├── 461807142278350376_1_461807142277950370_103
│           │   │   ├── _cached_nodes.bin
│           │   │   ├── _disk.index
│           │   │   ├── _disk.index_max_base_norm.bin
│           │   │   ├── _pq_compressed.bin
│           │   │   ├── _pq_pivots.bin
│           │   │   ├── _pq_pivots.bin_centroid.bin
│           │   │   ├── _pq_pivots.bin_chunk_offsets.bin
│           │   │   ├── _pq_pivots.bin_rearrangement_perm.bin
│           │   │   └── _sample_data.bin
│           │   └── 461807142278844734_1_461807142278494730_103
│           │       ├── _cached_nodes.bin
│           │       ├── _disk.index
│           │       ├── _disk.index_max_base_norm.bin
│           │       ├── _pq_compressed.bin
│           │       ├── _pq_pivots.bin
│           │       ├── _pq_pivots.bin_centroid.bin
│           │       ├── _pq_pivots.bin_chunk_offsets.bin
│           │       ├── _pq_pivots.bin_rearrangement_perm.bin
│           │       └── _sample_data.bin
│           ├── json_stats
│           │   ├── 461807142127582571_1_461807142127352192_104
│           │   │   └── shared_key_index
│           │   │       ├── 8e54b2e6546a4496b49dc0e375214da8.fast
│           │   │       ├── 8e54b2e6546a4496b49dc0e375214da8.fieldnorm
│           │   │       ├── 8e54b2e6546a4496b49dc0e375214da8.idx
│           │   │       ├── 8e54b2e6546a4496b49dc0e375214da8.pos
│           │   │       ├── 8e54b2e6546a4496b49dc0e375214da8.store
│           │   │       ├── 8e54b2e6546a4496b49dc0e375214da8.term
│           │   │       └── meta.json
│           │   ├── 461807142136917958_1_461807142136547575_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142146553490_1_461807142146183099_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142156099185_1_461807142155728790_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142165474847_1_461807142165104451_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142174761117_1_461807142174390728_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142184218762_1_461807142183848361_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142193716751_1_461807142193346360_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142203516608_1_461807142203146227_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142212766779_1_461807142212396396_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142222578803_1_461807142222348422_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142232077968_1_461807142231707587_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142241930485_1_461807142241700114_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142251499934_1_461807142251269570_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142266497459_1_461807142266127078_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142276759363_1_461807142276388982_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   ├── 461807142278350652_1_461807142277950370_104
│           │   │   └── shared_key_index
│           │   │       └── meta.json
│           │   └── 461807142278845048_1_461807142278494730_104
│           │       └── shared_key_index
│           │           └── meta.json
│           ├── raw_datas
│           └── tmp
│               └── json_stats
└── mmap
    ├── 461807142113844434
    │   └── 104
    │       ├── 2000
    │       ├── 2001
    │       ├── 2002
    │       ├── 2003
    │       ├── 2004
    │       ├── 2005
    │       ├── 2006
    │       ├── 2007
    │       ├── 2008
    │       └── 2009
...(Many folders with similar contents to folder 461807142113844434)
    ├── 461807142127352192
    │   └── 104
    │       ├── 2000
    │       │   ├── 0
    │       │   ├── 1
    │       │   ├── 10
    │       │   ├── 100
    │       │   ├── 101
    │       │   ├── 102
...(Many folders with similar contents to folder 461807142127352192)

**Reply by foxspy** [reply]
It looks like these indexes are being loaded normally; the _disk.index part will not be loaded into memory. During search, access to this data is done only through AIO, so you may want to double-check that.

**Reply by Buddingpopp** [reply]
Do you mean that even though I used Discann indexes, my indexes were still loaded into memory?
